File: GEMM/Python/GeneratedCode.py
# ===================================================================================================================
#               numpy Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Python numpy general matrix-matrix multiply def."

# SUGGESTION 1 # ----------------------------------

import numpy as np

def matrix_multiply(A, B):
    return np.dot(A, B)


# SUGGESTION 2 # ----------------------------------

import numpy as np

def matrix_multiply(A, B):
    return np.lib.arraysetops.matmat(A, B)


# SUGGESTION 3 # ----------------------------------

import numpy as np

def matrix_multiply(A, B):
    result = np.zeros((A.shape[0], B.shape[1]))
    for i in range(A.shape[0]):
        for j in range(B.shape[1]):
            for k in range(A.shape[1]):
                result[i, j] += A[i, k] * B[k, j]
    return result

# ===================================================================================================================
#               cuPy Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Python cuPy general matrix-matrix multiply def."

# SUGGESTION 1 # ----------------------------------

import cupy

def matrix_multiply(A, B):
    return cupy.mm(A, B)


# SUGGESTION 2 # ----------------------------------

import cupy

def matrix_multiply(A, B):
    return cupy.mm(A, B, stride=1)


# SUGGESTION 3 # ----------------------------------

import cupy

def matrix_multiply(A, B):
    block_size = 16
    num_blocks = (A.shape[0] + block_size - 1) // block_size
    result = cupy.zeros((A.shape[0], B.shape[1]), dtype=A.dtype)
    for i in range(num_blocks):
        for j in range(num_blocks):
            block_a = A[i*block_size:(i+1)*block_size, :]
            block_b = B[:, j*block_size:(j+1)*block_size]
            result[i*block_size:(i+1)*block_size, j*block_size:(j+1)*block_size] = \
                cupy.mm(block_a, block_b)
    return result

# ===================================================================================================================
#               pyCUDA Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Python pyCUDA general matrix-matrix multiply def."

# SUGGESTION 1 # ----------------------------------

import pycuda.autoinit
from pycuda import driver

def matrix_multiply(A, B):
    # Create GPU buffers for the matrices
    A_buf = driver.create_buffer(A.shape, A.dtype)
    B_buf = driver.create_buffer(B.shape, B.dtype)

    # Copy the data from the CPU arrays to the GPU buffers
    driver.copy_to_device(A, A_buf)
    driver.copy_to_device(B, B_buf)

    # Define the kernel function
    @driver.register_kernel
    def matmul(A, B):
        # Perform matrix multiplication
        C = A @ B
        return C

    # Execute the kernel on the GPU
    C_buf = driver.empty_like(A_buf)
    matmul[A_buf.shape[0], A_buf.shape[1], B_buf.shape[1]](A_buf, B_buf, C_buf)

    # Copy the result from the GPU buffer to a CPU array
    C = driver.copy_to_host(C_buf)

    return C


# SUGGESTION 2 # ----------------------------------

import pycuda.autoinit
from pycuda import driver

def matrix_multiply(A, B):
    # Create a shared memory pool
    smem = driver.shared_memory_pool(A.shape[0], B.shape[1])

    # Allocate space for the matrices in the shared memory pool
    A_smem = smem.allocate(A.shape[0], A.shape[1])
    B_smem = smem.allocate(B.shape[0], B.shape[1])

    # Copy the data from the CPU arrays to the shared memory
    driver.copy_to_device(A, A_smem)
    driver.copy_to_device(B, B_smem)

    # Define the kernel function
    @driver.register_kernel
    def matmul(A, B):
        # Perform matrix multiplication using shared memory
        C = A @ B
        return C

    # Execute the kernel on the GPU
    C_smem = smem.allocate(A.shape[0], B.shape[1])
    matmul[A_smem.shape[0], A_smem.shape[1], B_smem.shape[1]](A_smem, B_smem, C_smem)

    # Copy the result from the shared memory to a CPU array
    C = driver.copy_to_host(C_smem)

    return C


# SUGGESTION 3 # ----------------------------------

import pycuda.autoinit
from pycuda import driver

def matrix_multiply(A, B):
    # Create a shared memory pool
    smem = driver.shared_memory_pool()

    # Allocate space for the matrices in the shared memory pool dynamically
    A_smem = smem.dynamic_allocate(A.shape[0], A.shape[1])
    B_smem = smem.dynamic_allocate(B.shape[0], B.shape[1])

    # Copy the data from the CPU arrays to the shared memory
    driver.copy_to_device(A, A_smem)
    driver.copy_to_device(B, B_smem)

    # Define the kernel function
    @driver.register_kernel
    def matmul(A, B):
        # Perform matrix multiplication using shared memory
        C = A @ B
        return C

    # Execute the kernel on the GPU
    C_smem = smem.dynamic_allocate(A.shape[0], B.shape[1])
    matmul[A_smem.shape[0], A_smem.shape[1], B_smem.shape[1]](A_smem, B_smem, C_smem)

    # Copy the result from the shared memory to a CPU array
    C = driver.copy_to_host(C_smem)

    return C

# ===================================================================================================================
#               Numba Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Python Numba general matrix-matrix multiply def."

# SUGGESTION 1 # ----------------------------------

import numba

@numba.jit
def matrix_multiply(A, B):
    """
    Multiplies two matrices A and B and returns the result.
    """
    n = A.shape[0]
    m = B.shape[1]
    C = np.zeros((n, m))
    for i in range(n):
        for j in range(m):
            for k in range(A.shape[1]):
                C[i, j] += A[i, k] * B[k, j]
    return C


# SUGGESTION 2 # ----------------------------------

import numba
import numpy as np

@numba.jit
def strided_matrix_multiply(A, B, stride_a, stride_b):
    """
    Multiplies two matrices A and B with strides stride_a and stride_b.
    """
    n = A.shape[0]
    m = B.shape[1]
    C = np.zeros((n, m))
    for i in range(n):
        for j in range(m):
            for k in range(A.shape[1]):
                C[i, j] += A[i*stride_a, k] * B[k*stride_b, j]
    return C


# SUGGESTION 3 # ----------------------------------

import numba
import numpy as np

@numba.jit
def matrix_multiply_broadcast(A, B, broadcast_dim=None):
    """
    Multiplies two matrices A and B and returns the result.
    If broadcast_dim is None, it will automatically determine the broadcast dimension.
    """
    if broadcast_dim is None:
        broadcast_dim = max(A.ndim, B.ndim) - 1
    
    C = np.zeros((A.shape[:broadcast_dim] + B.shape[broadcast_dim:]))
    for i in range(A.shape[0]):
        for j in range(B.shape[0]):
            for k in range(A.shape[1]):
                C[i, j, k] = A[i, k] * B[k, j]
    return C

# ------------------------------------------------------------------------------------------------------------------------------------------------------

File: GEMM/Julia/GeneratedCode.jl
# ===================================================================================================================
#               Threads Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia Threads general matrix-matrix multiply."

# SUGGESTION 1 # ----------------------------------

using Threads

# Create two matrices
A = Matrix(rand(Float64, 10, 10))
B = Matrix(rand(Float64, 10, 10))

# Define a function for matrix multiplication
function matmult(A, B)
    return A * B
end

# Split the matrices into smaller chunks
n = 2
chunksize = (size(A, 1) + n - 1) / n
A_chunks = collect(nothing, i -> view(A, :, i .* chunksize .+ (i == div(n, 2) ? 1 : 0)))
B_chunks = collect(nothing, i -> view(B, :, i .* chunksize .+ (i == div(n, 2) ? 1 : 0)))

# Perform matrix multiplication in parallel
@threads for i in 1:n for j in 1:n
    local chunk_a = A_chunks[i]
    local chunk_b = B_chunks[j]
    local result = matmult(chunk_a, chunk_b)
    # Write the result back to the main array
    view(A, :, (i - 1) * chunksize .+ 1 : min(i * chunksize, size(A, 1))) .= result
end


# SUGGESTION 2 # ----------------------------------

using Distributed

# Create two matrices
A = Matrix(rand(Float64, 10, 10))
B = Matrix(rand(Float64, 10, 10))

# Define a function for matrix multiplication
function matmult(A, B)
    return A * B
end

# Split the matrices into smaller chunks
n = 2
chunksize = (size(A, 1) + n - 1) / n
A_chunks = collect(nothing, i -> view(A, :, i .* chunksize .+ (i == div(n, 2) ? 1 : 0)))
B_chunks = collect(nothing, i -> view(B, :, i .* chunksize .+ (i == div(n, 2) ? 1 : 0)))

# Create a distributed array to store the results
result = Dict()

# Perform matrix multiplication in parallel
for i in 1:n for j in 1:n
    local chunk_a = A_chunks[i]
    local chunk_b = B_chunks[j]
    local result = matmult(chunk_a, chunk_b)
    # Store the result in the distributed array
    result[(i - 1) * chunksize .+ 1 : min(i * chunksize, size(A, 1))] = result
end

# Collect the results from the distributed array
final_result = collect(result)


# SUGGESTION 3 # ----------------------------------

using Parallel

# Create two matrices
A = Matrix(rand(Float64, 10, 10))
B = Matrix(rand(Float64, 10, 10))

# Define a function for matrix multiplication
function matmult(A, B)
    return A * B
end

# Split the matrices into smaller chunks
n = 2
chunksize = (size(A, 1) + n - 1) / n
A_chunks = collect(nothing, i -> view(A, :, i .* chunksize .+ (i == div(n, 2) ? 1 : 0)))
B_chunks = collect(nothing, i -> view(B, :, i .* chunksize .+ (i == div(n, 2) ? 1 : 0)))

# Perform matrix multiplication in parallel
pmap(matmult, (A_chunks, B_chunks), axis=1) do (chunk_a, chunk_b)
    return chunk_a * chunk_b
end

# Convert the result back to a Matrix
final_result = convert(Matrix, final_result)



# ===================================================================================================================
#               CUDA Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia CUDA general matrix-matrix multiply."

# SUGGESTION 1 # ----------------------------------

using CUDA

# Define the matrices
A = CuMatrix(rand(Float32, 10, 10))
B = CuMatrix(rand(Float32, 10, 10))

# Perform matrix-matrix multiplication
C = A * B

# Print the result
println(C)


# SUGGESTION 2 # ----------------------------------

using CUDA

# Define the matrices
A = CuMatrix(rand(Float32, 10, 10))
B = CuMatrix(rand(Float32, 10, 10))

# Define strides for A and B
stride_a = (2, 3)
stride_b = (3, 2)

# Perform strided matrix-matrix multiplication
C = A[:, stride_a] * B[stride_b, :]

# Print the result
println(C)


# SUGGESTION 3 # ----------------------------------

using CUDA

# Define the matrices
A = CuMatrix(rand(Float32, 10, 10))
B = CuMatrix(rand(Float32, 10, 10))

# Define shared memory blocks
block_size = 32
shared_memory = CuArray{Float32}(undef, block_size, block_size)

# Perform shared memory matrix-matrix multiplication
@cuda threads per block = block_size
syncthreads()

for i = 1:div(size(A, 1), block_size)
    for j = 1:div(size(A, 2), block_size)
        @shared begin
            local A_local = A[i:min(i + block_size - 1, size(A, 1)), j:min(j + block_size - 1, size(A, 2))]
            local B_local = B[i:min(i + block_size - 1, size(B, 1)), j:min(j + block_size - 1, size(B, 2))]
            local C_local = zeros(Float32, block_size, block_size)
            for k = 1:min(block_size, size(A, 1))
                for l = 1:min(block_size, size(A, 2))
                    C_local[k, l] = dot(A_local[k, l], B_local[l, k])
                end
            end
            syncthreads()
            if ThreadsPerBlock() > 1
                shared_memory[threadIdx().y * block_size + threadIdx().x] = C_local
            end
        end
    end
end

# Copy the result from shared memory to global memory
C = Array( Float32, size(A, 1), size(A, 2) )
copy!(C, shared_memory)

# Print the result
println(C)



# ===================================================================================================================
#               AMDGPU Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia AMDGPU general matrix-matrix multiply."

# SUGGESTION 1 # ----------------------------------

using AMDGPU

# Define the matrices
A = AMDGPU.Matrix{Float32, 2}(rand(10, 10), rand(10, 10))
B = AMDGPU.Matrix{Float32, 2}(rand(10, 10), rand(10, 10))

# Perform matrix-matrix multiplication
C = A * B

# Print the result
println(C)


# SUGGESTION 2 # ----------------------------------

using AMDGPU

# Define the matrices
A = AMDGPU.Matrix{Float32, 2}(rand(10, 10), rand(10, 10))
B = AMDGPU.Matrix{Float32, 2}(rand(10, 10), rand(10, 10))

# Define the strides
strideA = (2, 3)
strideB = (3, 2)

# Perform strided matrix-matrix multiplication
C = A * B'::StridedMatrixMultiplication(strideA, strideB)

# Print the result
println(C)


# SUGGESTION 3 # ----------------------------------

using AMDGPU

# Define the matrices
A = AMDGPU.Matrix{Float32, 2}(rand(10, 10), rand(10, 10))
B = AMDGPU.Matrix{Float32, 2}(rand(10, 10), rand(10, 10))

# Define the broadcasting shape
broadcast_shape = (10, 10, 1, 1)

# Perform matrix-matrix multiplication with broadcasting
C = A * B[:, :, 1, 1]::BroadcastedMatrixMultiplication(broadcast_shape)

# Print the result
println(C)



# ===================================================================================================================
#               KernelAbstractions Programming Model
# ===================================================================================================================

# Prompt: "Create 3 code suggestions using the following parameters: Julia KernelAbstractions general matrix-matrix multiply."

# SUGGESTION 1 # ----------------------------------

using KernelAbstractions

# Define the matrices
A = KernelAbstractions.Matrix{Float64, 2}(rand(10, 10), rand(10, 10))
B = KernelAbstractions.Matrix{Float64, 2}(rand(10, 10), rand(10, 10))

# Perform matrix-matrix multiplication
C = A * B

# Print the result
println(C)


# SUGGESTION 2 # ----------------------------------

using KernelAbstractions

# Define the matrices
A = KernelAbstractions.Matrix{Float64, 2}(rand(10, 10), rand(10, 10))
B = KernelAbstractions.Matrix{Float64, 2}(rand(10, 10), rand(10, 10))

# Define the strides
strideA = (2, 3)
strideB = (3, 2)

# Perform strided matrix-matrix multiplication
C = A * B'::StridedMatrixMultiplication(strideA, strideB)

# Print the result
println(C)


# SUGGESTION 3 # ----------------------------------

using KernelAbstractions

# Define the matrices
A = KernelAbstractions.Matrix{Float64, 2}(rand(10, 10), rand(10, 10))
B = KernelAbstractions.Matrix{Float64, 2}(rand(10, 10), rand(10, 10))

# Define the broadcasting shape
broadcast_shape = (10, 10, 1, 1)

# Perform matrix-matrix multiplication with broadcasting
C = A * B[:, :, 1, 1]::BroadcastedMatrixMultiplication(broadcast_shape)

# Print the result
println(C)



# ---------------------------------------------------------------------------------------------------------------------------------------------------------

